File: learn/cpp/classes.cpp
#include <iostream>

using namespace std;

class Foo {
public:
	explicit Foo(int arg = 42)
		: a(arg)
	{
		cout << "Foo's constructor, a = " << a << endl;
	};

private:
	int a;
};

class Parent {
public:
	explicit Parent(int arg = 0)
		: a(arg)
	{
		cout << "Parent's constructor" << endl;
	}

	//void hello() { cout << "This is parent." << endl; };
	virtual void hello() { cout << "This is parent." << endl; };

	int a = 0;
};

class Child : public Parent {
public:
	explicit Child(int arg = 0)
		//: Parent(arg)
		: a(initFoo)
	{
		cout << "Child's constructor" << endl;
	};

	void hello() { cout << "This is child." << endl; };
	//virtual hello() { cout << "This is parent." << endl; };

	const int initFoo = 420;
	Foo a;
};

int main()
{
	Parent *p;
	Child c(3);
	p = &c;

	c.hello();
	p->hello();
}

File: learn/java/classes.java
class wrap {
public static void main(String[] args) {

class Parent {
	private int a = 50;
	public Parent(int a) {
		System.out.println("Parent's constructor");
		System.out.println(this.a);
	}
}

class Child extends Parent {
	public int a = 40;
	public Child(int a) {
		super(a);
		System.out.println("Child's constructor");
		System.out.println(a);
	}
}

Child c = new Child(30);
System.out.println(c.a);

}
}
